curly/upload/request.go

```diff
@@ -0,0 +1,53 @@
+package upload
+
+import (
+	"compress/gzip"
+	"fmt"
+	"io"
+	"log"
+	"mime/multipart"
+	"net/http"
+)
+
+type Uploader struct {
+	c         *http.Client
+	uploadURL string
+}
+
+func New(c *http.Client, uploadURL string) *Uploader {
+	return &Uploader{
+		c:         c,
+		uploadURL: uploadURL,
+	}
+}
+
+func (u *Uploader) Upload(r io.Reader, filename string) (*http.Request, error) {
+
+	rBody, wPipe := io.Pipe()
+	writer := multipart.NewWriter(wPipe)
+	go func() {
+		defer wPipe.Close()
+		defer writer.Close()
+
+		gzfilename := fmt.Sprintf("%s.gz", filename)
+		part, err := writer.CreateFormFile("file", gzfilename)
+
+		gzipW := gzip.NewWriter(part)
+		defer gzipW.Close()
+
+		_, err = io.Copy(gzipW, r)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	req, err := http.NewRequest(http.MethodPost, u.uploadURL, rBody)
+	if err != nil {
+		return nil, err
```

Comment:
if reading rBody failes here or on c.Do(req) - maybe there is no point copying data to wPipe and goroutine is doing not needed work?

for testing this - put time.Sleep(10 minutes) before log.Fatal on code block
```
		resp, err := c.Do(req)
		if err != nil {
			log.Fatal(err)
		}
```		
and log when io.Copy above finishes working. As parameters pass some huge file to download and -uploadurl some random url that doesn't accepts uploads
 
 pipeReader and pipeWriter have wonderful methods CloseWithError